pyquil/quil.py

```diff
@@ -209,13 +222,97 @@ def if_then(self, classical_reg, if_program, else_program=None):
         :returns: The Quil Program with the branching instructions added.
         :rtype: Program
         """
-
         else_program = else_program if else_program is not None else Program()
 
-        branch = If(Addr(classical_reg))
-        branch.Then.inst(if_program)
-        branch.Else.inst(else_program)
-        return self.inst(branch)
+        # IF [c]:
+        #    instrA...
+        # ELSE:
+        #    instrB...
+        #
+        # =>
+        #
+        #   JUMP-WHEN @THEN [c]
+        #   instrB...
+        #   JUMP @END
+        #   LABEL @THEN
+        #   instrA...
+        #   LABEL @END
+        label_then = LabelPlaceholder("THEN")
+        label_end = LabelPlaceholder("END")
+        self.inst(JumpWhen(target=label_then, condition=Addr(classical_reg)))
+        self.inst(else_program)
+        self.inst(Jump(label_end))
+        self.inst(JumpTarget(label_then))
+        self.inst(if_program)
+        self.inst(JumpTarget(label_end))
+        return self
+
+    def alloc(self):
+        """
+        Get a new qubit.
+
+        :return: A qubit.
+        :rtype: Qubit
+        """
+        return QubitPlaceholder()
+
+    def out(self):
+        """
+        Converts the Quil program to a readable string.
+
+        :return: String form of a program
+        :rtype: string
+        """
+        s = ""
+        for dg in self._defined_gates:
+            s += dg.out()
+            s += "\n"
+        for instr in self.instructions:
+            s += instr.out() + "\n"
+        return s
+
+    def get_qubits(self):
+        """
+        Returns all of the qubit indices used in this program, including gate applications and
+        allocated qubits. e.g.
+            >>> p = Program()
+            >>> p.inst(("H", 1))
+            >>> p.get_qubits()
+            {1}
+            >>> q = p.alloc()
+            >>> len(p.get_qubits())
+            2
+
+        :return: A set of all the qubit indices used in this program
+        :rtype: set
+        """
+        qubits = set()
+        for instr in self.instructions:
+            if isinstance(instr, Gate):
+                qubits |= {q.index for q in instr.qubits}
+            elif isinstance(instr, Measurement):
+                qubits.add(instr.qubit.index)
+        return qubits
+
+    def is_protoquil(self):
+        """
+        Protoquil programs may only contain gates, no classical instructions and no jumps.
```

Comment:
One weird thing about this, I think, is the following:
```
>>> p = Program()
>>> p.inst(("MEASURE", 0))
<pyquil.quil.Program object at 0x7f5888691b90>
>>> p.is_protoquil()
True
>>> p.inst(MEASURE(0))
<pyquil.quil.Program object at 0x7f5888691b90>
>>> p.is_protoquil()
False
```
So strictly speaking, it's wrong. Right?

Reply:
Made the following changes:
- inst now specifically looks for `MEASURE` as a keyword
- Gate will validate that the gate name is not a reserved word
- passing in a raw string like `JUMP @LABEL` will first get parsed by the parser